### PR TITLE
[video_player] Fix iOS crash with multiple players

### PR DESCRIPTION
**Pull request by turekj**
This PR fixes crash in `video_player` package on iOS. 

https://github.com/flutter/flutter/issues/124937

### Detailed description

I can observe the crash when displaying a couple of the different players (different URLs) inside a `ListView`. The crash happens inside of `AVFoundation` framework:

```objc
[AVPlayer _createAndConfigureFigPlayerWithType:completionHandler:]
```

In order to debug the issue, I ran the application using the plugin with `Zombie Objects` inspection turned on. The `Zombie Objects` reports the following issue:

```
*** -[FLTVideoPlayer retainWeakReference]: message sent to deallocated instance 0x6030009b2e10
```

This, in conjunction with the `NSKeyValueWillChange` line present in the stack trace led me to believe, that culprit is sending a KVO notification to the `FLTVideoPlayer` instance that's deallocated.

Next, I examined the plugin code and identified one property that doesn't have the KVO removed. In `addObserversForItem:player:` method in `FLTVideoPlayerPlugin.m`:

```
  [player addObserver:self
           forKeyPath:@"rate"
              options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
              context:rateContext];
```

The observer for `@"rate"` is never cleaned up. To be entirely sure that the issue comes from KVO that's not cleaned up, I've added the following class:

```
@implementation EmptyObserver

- (void)observeValueForKeyPath:(NSString *)path
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {}

@end
```

and registered the observation as follows (notice that `EmptyObserver` is never retained and deallocated instantly):

```
  [player addObserver:[[EmptyObserver alloc] init]
           forKeyPath:@"rate"
              options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
              context:rateContext];
```

The exception I got seems to be matching the underlying issue:

```
*** -[EmptyObserver retainWeakReference]: message sent to deallocated instance 0x6020001ceb70
```

This means the fix for the issue is to add the following to `disposeSansEventChannel` method:

```
[self.player removeObserver:self forKeyPath:@"rate"];
```

After applying the patch, I can no longer crash the player.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m [hunk 1]
```diff
@@ -530,8 +530,11 @@ - (void)disposeSansEventChannel {
   [currentItem removeObserver:self forKeyPath:@"playbackLikelyToKeepUp"];
   [currentItem removeObserver:self forKeyPath:@"playbackBufferEmpty"];
   [currentItem removeObserver:self forKeyPath:@"playbackBufferFull"];
+  [self.player removeObserver:self forKeyPath:@"rate"];
 
   [self.player replaceCurrentItemWithPlayerItem:nil];
+  _player = nil;
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:536]: why is this required?

**Review by hellohuanlin — CHANGES_REQUESTED** [review]
can you explain in the PR description why the crash happens and how this fixes the crash?

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m (hunk 1 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:536]: When applied the original fix, the application using the video_player started crashing while doing the Hot Restart with Flutter attached in debug mode. The reason for this crash is that the player is no longer KVO subscribed to property rate, therefore remove observer calls crashes.

With additional logging I was able to confirm that disposeSansEvent method is called twice on the same player. The issue is not present in the original code, because of replace current item with nil call that happens in the dispose (the line before I nullify the player). For the subsequent call, the player item is nullified, therefore remove observer call is no-op. I applied the same style of the solution to the player.

Comment (hellohuanlin) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:536]: when it's called twice, should we short circuit up front? 

```
- (void)dispose... {
  if (!_player) {
    return;
  } 
  ... 
}
```

Comment (turekj) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:536]: I'll short circuit it with:

```objc
- (void)disposeSansEventChannel {
  if (_disposed) {
    return;
  }
  _disposed = YES;
// ...
}
```

as `_disposed` is already set there.

Comment (hellohuanlin) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:536]: makes sense. Can you also add some comments here?

**Comment by hellohuanlin** [comment]
> I'll update the code as requested and provide some additional comments.

Oh i meant comment in the code. It would be helpful for future readers.

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m (hunk 1 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m]: @hellohuanlin I've updated the code according to our discussion. I've added a comment on the short circuit method.

I don't think removing the observer requires additional comments as the API specifies to clear any observations you've subscribed to in the first place. LMK if you'd add anything else, please 🙏

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m [hunk 2]
```diff
@@ -519,6 +519,14 @@ - (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments
 /// is useful for the case where the Engine is in the process of deconstruction
 /// so the channel is going to die or is already dead.
 - (void)disposeSansEventChannel {
+  // This check prevents the crash caused by removing the KVO observers twice.
+  // When performing a Hot Restart, the leftover players are disposed directly
+  // and also receive onTextureUnregistered: callback leading to possible
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:524]: nit: just wanna be clearer

"the leftover players are disposed twice - once directly (from which function?), and once when receiving onTextureUnregistered callback"

**Comment by stuartmorgan** [comment]
> * [ ]  I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
> * [ ]  I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
> ...
> * [ ]  I added new tests to check the change I am making, or this PR is [test-exempt].

@turekj Could you update the PR to add these missing components? All of these elements are required before the PR can be fully reviewed and landed.

**Comment by stuartmorgan** [comment]
From PR triage: @turekj Are you still planning on updating this per the feedback above?

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m (hunk 2 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m]: @hellohuanlin I've updated the comment yet again to make it clear

**Comment by turekj** [comment]
@hellohuanlin I've updated the comments so that they're clearer ✅ 

@stuartmorgan I've added a changelog entry + bumped the package number. 

I'll take a look into adding a test case for this issue.

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m [hunk 3]
```diff
@@ -396,6 +398,111 @@ - (void)testSeekToleranceWhenSeekingToEnd {
...
+    player = videoPlayerPlugin.playersByTextureId[textureMessage.textureId];
+    XCTAssertNotNil(player);
+    avPlayer = player.player;
+
+    [videoPlayerPlugin dispose:textureMessage error:&error];
+    XCTAssertNil(error);
+  }
+
+  // [FLTVideoPlayerPlugin dispose:error:] selector is dispatching the [FLTVideoPlayer dispose] call
+  // with a 1-second delay keeping a strong reference to the player. The polling ensures the player
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:442]: what does this 1 sec delay come from?

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m [hunk 4]
```diff
@@ -439,4 +546,22 @@ - (void)validateTransformFixForOrientation:(UIImageOrientation)orientation {
...
+  do {
+    if (condition()) {
+      return;
+    }
+
+    [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.002]];
+  } while ([[NSDate date] timeIntervalSinceNow] < [end timeIntervalSinceNow]);
+
+  if (condition()) {
+    return;
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:561]: is this check necessary?

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m [hunk 5]
```diff
@@ -439,4 +546,22 @@ - (void)validateTransformFixForOrientation:(UIImageOrientation)orientation {
   XCTAssertEqual(t.ty, expectY);
 }
 
+- (void)waitForCondition:(BOOL (^)(void))condition withTimeout:(NSTimeInterval)timeout {
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:549]: i think you can just use NSPredicate and expectation for this kind of stuff

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m [hunk 6]
```diff
@@ -519,6 +519,14 @@ - (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments
 /// is useful for the case where the Engine is in the process of deconstruction
 /// so the channel is going to die or is already dead.
 - (void)disposeSansEventChannel {
+  // This check prevents the crash caused by removing the KVO observers twice.
+  // When performing a Hot Restart, the leftover players are disposed directly
+  // by [FLTVideoPlayerPlugin initialize:] method and also receive
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m:524]: nit: "the leftover players are disposed once directly by ... method and then disposed again when receiving ..."

**Review by hellohuanlin — CHANGES_REQUESTED** [review]
Some feedbacks mainly on unit tests

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m (hunk 3 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:442]: The `[FLTVideoPlayerPlugin dispose:error:]` has this piece of code:

```
dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)),
               dispatch_get_main_queue(), ^{
                 if (!player.disposed) {
                   [player dispose];
                 }
               });
```

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m (hunk 5 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m]: Great idea, I've refactored the code ✅

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m (hunk 4 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m]: Not valid anymore

File: packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m (hunk 6 above)
Comment (turekj) [reply packages/video_player/video_player_avfoundation/ios/Classes/FLTVideoPlayerPlugin.m]: Updated the wording ✅

**Comment by turekj** [comment]
@hellohuanlin implemented all rounds of feedback ✅

**Review by hellohuanlin — APPROVED** [review]

**Review by tarrinneal — APPROVED** [review]

**Comment by stuartmorgan** [comment]
@turekj Could you resolve the conflicts that have shown up? Once that's done this can be landed.

**Comment by turekj** [comment]
@stuartmorgan conflicts resolved ✅

**Comment by stuartmorgan** [comment]
Thanks!
